### PR TITLE
 Update non-thread-safe data on the main thread

### PR DESCRIPTION
**Pull request by dimitry-**
For global maps which are currently used from the main thread
    we should make sure that during capturing they are updated on
    the main thread as well, to avoid undefined behavior.
    
    This change moves updates of capture data to the main thread.
    It also introduces OnCaptureStarted and OnCaptureComplete
    callbacks to the listener.

### CODE REVIEW

**Comment by pierricgimmig** [comment]
This will have big performance implication and slow down the application quite a bit when streaming in timers. It's always been the goal to process timers and events as fast as possible without blocking the main thread. Through the use of the blockchain data structure, which never migrates elements, a worker thread can append timers while the rendering/main thread displays them without locking.  We are talking 100K timers per second here, scheduling each of them individually seems overkill to me.

> Since we are using global variables to store capture data, we should make sure that access to that data is happening on the single thread. 

I don't see how the presence of global variables entails the single threaded approach. Also, I think we should move away from globals.

**Comment by dimitry-** [comment]
I agree that we need to move away from globals, but this is much larger scope - I am currently trying to fix races in capture process. 

For context: CaptureStarted/Complete are introduced to start solving races around Capture::State in order to do that I am moving lifecycle inside CaptureClient and this is the first step for this process.

I moved thread-safe operations back to capture thread, this requires CaptureStarted to block the capture thread until initialization is done on the main thread. ptal.

File: OrbitGl/App.cpp [hunk 1]
```diff
@@ -111,6 +111,52 @@ void OrbitApp::SetCommandLineArguments(const std::vector<std::string>& a_Args) {
   }
 }
 
+static bool IsTrue(bool* flag) { return *flag; }
```
Comment (dpallotti) [inline OrbitGl/App.cpp:114]: I would prefer the use of lambdas in `absl::Condition`, to avoid these very generic tiny static functions.

File: OrbitGl/App.cpp [hunk 2]
```diff
@@ -120,35 +166,47 @@ void OrbitApp::OnKeyAndString(uint64_t key, std::string str) {
 }
 
 void OrbitApp::OnCallstack(CallStack callstack) {
-  Capture::GSamplingProfiler->AddUniqueCallStack(callstack);
+  main_thread_executor_->Schedule([callstack = std::move(callstack)] {
```
Comment (dpallotti) [inline OrbitGl/App.cpp:169]: I have to agree that this seems like *a lot* of scheduling actions if a new one is scheduled for each event, even if this is not done for `Timer`s which are the most frequent ones. Not sure how a reasonably quick fix would look like though. But as this is not directly related to the issue with `CaptureClient` (right?), was this actually causing problems?

File: OrbitGl/App.cpp [hunk 3]
```diff
@@ -111,6 +111,52 @@ void OrbitApp::SetCommandLineArguments(const std::vector<std::string>& a_Args) {
   }
 }
 
+static bool IsTrue(bool* flag) { return *flag; }
+
+void OrbitApp::OnCaptureStarted() {
```
Comment (dpallotti) [inline OrbitGl/App.cpp:116]: So this is *not* scheduled on the main thread and has to wait for the scheduled action to complete on the main thread, right?

File: OrbitGl/App.cpp (hunk 2 above)
Comment (ronaldfw) [reply OrbitGl/App.cpp:169]: This should at least be measured to understand the performance impact, o/w it seems to be hard to make a decision here.

File: OrbitGl/App.cpp (hunk 3 above)
Comment (dimitry-) [reply OrbitGl/App.cpp:116]: right.

File: OrbitGl/App.cpp (hunk 2 above)
Comment (dimitry-) [reply OrbitGl/App.cpp:169]: I have not seen crashes because of this - but this is obviously not thread_safe so I think we need to fix it.

Comment (dimitry-) [reply OrbitGl/App.cpp:169]: ok, I made sampling profiler thread-safe, lets try again.

**Comment by dimitry-** [comment]
sorry, I had to force-push this - ptal.

File: OrbitGl/App.cpp (hunk 3 above)
Comment (pierricgimmig) [reply OrbitGl/App.cpp:114]: This I don't understand, the capture start is initiated from the UI, so on the main thread, shouldn't OnCaptureStarted() already be on the main thread? 

In OrbitApp::StartCapture(), on the main thread, we schedule the call to capture_client_->Capture() on a worker thread. We then in turn call "OnCaptureStarted" from the worker thread which needs to schedule some work on the main thread. Are those hops necessary? Why can't we call capture_client_->Capture() from the main thread?

Comment (dimitry-) [reply OrbitGl/App.cpp:114]: So the end-goal is to move state from Capture to CaptureClient, in order for this to be consitent it will need to be protected by mutex and we want to avoid making the mutex and implementation details for synchronization global. This is why I added CaptureStarted/Complete as part of Capture lifecycle. Note that this will also improve error handling, - for example in the case of unsuccessful capture you do not end up cleaning up previous results.

We do not call anything that involves blocking I/O on the main thread in order to avoid freezes to UI. CaptureClient::Capture calls streaming grpc Read which could block until next event has arrived.

Comment (pierricgimmig) [reply OrbitGl/App.cpp:114]: Ah yes, thanks for the precision.

File: OrbitCore/SamplingProfiler.cpp [hunk 4]
```diff
@@ -61,22 +61,22 @@ std::multimap<int, CallstackID> SamplingProfiler::GetCallstacksFromAddress(
 }
 
 //-----------------------------------------------------------------------------
-void SamplingProfiler::AddCallStack(CallstackEvent& callstack_event) {
+void SamplingProfiler::AddCallStack(CallstackEvent callstack_event) {
```
Comment (pierricgimmig) [inline OrbitCore/SamplingProfiler.cpp:64]: Why pass argument by value here? It seems we copy/move on the caller side, then again in this function.

Comment (dimitry-) [reply OrbitCore/SamplingProfiler.cpp:64]: reference leads to one extra-copy in this case: https://g3doc.corp.google.com/devtools/library_club/g3doc/totw/117.md?cl=head

Comment (pierricgimmig) [reply OrbitCore/SamplingProfiler.cpp:64]: @dimitry- , @dpallotti , this still doesn't make a whole lot of sense to me. If the object being passed by value has some POD data, then we end up copying this static data twice. Shouldn't AddCallstack take in a rvalue reference as argument? This way we save a copy if the caller std::moves the argument.  If not, then the caller has to create a temporary which makes the extra copy explicit.

Comment (dpallotti) [reply OrbitCore/SamplingProfiler.cpp:64]: You are correct about POD. Indeed, when the argument is fully POD you get "Clang-Tidy: std::move of the variable '...' of the trivially-constructible type '...' has no effect; remove std::move()".
But passing by const-ref or using the pass-by-valye-and-move pattern doesn't really make a significant difference with just a few POD fields. But you don't want many POD fields anyway, otherwise when you pass an r-value you are not able to move them, and be it that you copy once (passing an r-value reference you can't move from or a const-ref) or twice (passing by value), it's still too much copying. Instead you want to be able to move that data, by storing it inside the class with smart pointers, vectors, strings, etc, all of which can be moved. 
So the fact that you don't have too much POD data to copy regardless of whether you are copying or moving is pretty much an assumption. At this point, using pass-by-value-and-move is about adhering to a pattern that with a lot of movable data becomes quite relevant. If we always use it, then we also use it when we should use it.

You mention rvalue references. That would be the ideal parameter to pass r-values. But what if you want to pass an l-value (and make a copy of it)? Then you need the two overloads, one for the r-value, one for the l-value. That's an extra method and extra code. But what if you have two parameters (both of which to get into the class, are not just read)? Then you need four overloads, for the different combinations. And so on, that's not sustainable.
So what you do instead is have the parameter(s) by value and move from them. If you pass an l-value, the copy happens when you pass, and then you have the move in the method&mdash;if you were passing by const-ref, you would not have the move but you would still have the copy in the method. But if you pass an r-value, you have two moves and no copies&mdash;if you were passing by const-ref, you would not have the first move but then you would have to copy in the method, as you obviously can't move from a const-ref.
So l-value: one copy + one move instead of one copy. But r-value: **two moves instead of one copy**.

Comment (pierricgimmig) [reply OrbitCore/SamplingProfiler.cpp:72]: > You mention rvalue references. That would be the ideal parameter to pass r-values. But what if you want to pass an l-value (and make a copy of it)? Then you need the two overloads, one for the r-value, one for the l-value.

Well in that case you can create a temporary object at the call site without having to declare an extra method.

> **the fact that the move is cheap is an assumption**.
This is exactly my point, yes. You have to know about the type you're passing, in some cases there might be expensive copies of static data.

> So l-value: one copy + one move instead of one copy. But r-value: **two moves instead of one copy**.

But if you have a r-value as argument, you have only one move.

See https://repl.it/repls/UltimateOrangeredWebpages#main.cpp

File: OrbitGl/App.h [hunk 5]
```diff
@@ -302,6 +303,10 @@ class OrbitApp final : public DataViewFactory, public CaptureListener {
   const SymbolHelper symbol_helper_;
 
   std::unique_ptr<FramePointerValidatorClient> frame_pointer_validator_client_;
+
+  // Temporary objects used by CaptureListener implementation
+  std::unordered_map<uint64_t, orbit_client_protos::LinuxAddressInfo>
```
Comment (pierricgimmig) [inline OrbitGl/App.h:308]: Should this be a absl::flat_hash_map?

**Comment by dimitry-** [comment]
> Hi Dimitry, LGTM, thanks for getting rid of the scheduling of address_infos. The only thing is that callstack_event argument being passed by value and incurring one extra move. Let's discuss it offline.

Submitted this one, created #945 as a foolow up, we can continue discussion there.

File: OrbitCore/SamplingProfiler.cpp (hunk 4 above)
Comment (dpallotti) [reply OrbitCore/SamplingProfiler.cpp:72]: > Well in that case you can create a temporary object at the call site without having to declare an extra method.

That gets very annoying very quickly in general.

> You have to know about the type you're passing, in some cases there might be expensive copies of static data.

Is there expensive that to pass in this case though? This is a proto. I would guess moves are cheap.
